scipy/spatial/transform/rotation.py

```diff
@@ -0,0 +1,754 @@
+    quat[:, 3] = np.cos(angles / 2)
+    quat[:, AXIS_TO_IND[axis]] = np.sin(angles / 2)
+    return quat
+
+
+def _compose_quat(p, q):
+    product = np.empty((max(p.shape[0], q.shape[0]), 4))
+    # Scalar part of result
+    product[:, 3] = p[:, 3] * q[:, 3] - np.sum(p[:, :3] * q[:, :3], axis=1)
+    # Vector part of result
+    product[:, :3] = (p[:, None, 3] * q[:, :3] + q[:, None, 3] * p[:, :3] +
+                      np.cross(p[:, :3], q[:, :3]))
+    return product
+
+
+def _elementary_quat_compose(seq, angles, intrinsic=False):
+    # Initialize result to first axis
+    result = _make_elementary_quat(seq[0], angles[:, 0])
+
+    for idx, axis in enumerate(seq[1:], start=1):
+        if intrinsic:
+            result = _compose_quat(
+                result,
+                _make_elementary_quat(axis, angles[:, idx]))
+        else:
+            result = _compose_quat(
+                _make_elementary_quat(axis, angles[:, idx]),
+                result)
+    return result
+
+
+class Rotation(object):
+    """Rotation in 3 dimensions.
+
+    This class will include initializers from different representations,
+    converters and some useful algorithms such as quaternion slerp and
+    rotation estimation.
+
+    For initializing Rotations usage of `from_...` methods such as
+    `from_quaternion` is recommended instead of using `__init__`.
+
+    Methods
+    -------
+    from_quaternion
+    as_quaternion
+    from_dcm
+    as_dcm
+    from_rotvec
+    as_rotvec
+    from_euler
+    as_euler
+    inv
+    __mul__
+    apply
+    __getitem__
+    """
+    def __init__(self, quat, normalized=False, copy=True):
+        self._single = False
+        # Try to convert to numpy array
+        quat = np.asarray(quat, dtype=float)
+
+        if quat.ndim not in [1, 2] or quat.shape[-1] != 4:
+            raise ValueError("Expected `quat` to have shape (4,) or (N x 4), "
+                             "got {}.".format(quat.shape))
+
+        # If a single quaternion is given, convert it to a 2D 1 x 4 matrix but
+        # set self._single to True so that we can return appropriate objects
+        # in the `to_...` methods
+        if quat.shape == (4,):
+            quat = quat[None, :]
+            self._single = True
+
+        if normalized:
+            self._quat = quat.copy() if copy else quat
+        else:
+            self._quat = quat.copy()
+            # L2 norm of each row
+            norms = scipy.linalg.norm(quat, axis=1)
+
+            # Raise ValueError for zero (eps?) norm quaternions
+            zero_norms = norms == 0
+            if zero_norms.any():
+                raise ValueError("Found zero norm quaternions in `quat`.")
+
+            # Normalize each quaternion, ensuring norm is broadcasted along
+            # each column.
+            self._quat[~zero_norms] /= norms[~zero_norms][:, None]
+
+    @property
+    def n(self):
+        """Number of rotations contained in object."""
+        return self._quat.shape[0]
```

Comment:
I would just make this `def __len__` and users do `len(rots)` instead of `rots.n`